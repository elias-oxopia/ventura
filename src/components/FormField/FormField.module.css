.label {
  display: block;
  margin-bottom: 10px;
  font-weight: 500;
  font-size: 15px;
}

.errorMessage {
  display: block;
  margin-top: 5px;
  color: var(--color-error);
}

.hint {
  display: block;
  margin-top: 5px;
  color: var(--color-neutral-6);
}
