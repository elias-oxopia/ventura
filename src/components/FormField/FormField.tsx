import React, { ReactNode } from 'react';
import classNames from 'classnames';

import cssReset from '../../css-reset.module.css';
import styles from './FormField.module.css';

type Props = {
  label?: string;
  hint?: ReactNode;
  errorMessage?: string;
  className?: string;
  children: ReactNode;
  isSpaced?: boolean;
};

const FormField: React.FC<Props> = ({
  label,
  className,
  children,
  hint,
  errorMessage,
  isSpaced = false,
}: Props) => {
  const mergedClassNames = classNames(cssReset.ventura, className, {
    [styles.isSpaced]: isSpaced,
  });
  return (
    <div className={mergedClassNames}>
      {Boolean(label) && <span className={styles.label}>{label}</span>}
      {children}
      {Boolean(errorMessage) && <span className={styles.errorMessage}>{errorMessage}</span>}
      {Boolean(hint) && <div className={styles.hint}>{hint}</div>}
    </div>
  );
};

export default FormField;
