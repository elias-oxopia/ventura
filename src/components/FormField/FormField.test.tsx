import React from 'react';
import { render } from '@testing-library/react';

import FormField from './FormField';

describe('FormField', () => {
  test('snapshot', () => {
    const { asFragment } = render(<FormField>Form elements here</FormField>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('errorMessage should be rendered', () => {
    const { queryByText } = render(
      <FormField errorMessage="Something went wrong">Form elements here</FormField>,
    );
    expect(queryByText(/Something went wrong/)).not.toBe(null);
  });

  test('undefined errorMessage should not be rendered', () => {
    const { queryByText } = render(
      <FormField errorMessage={undefined}>Form elements here</FormField>,
    );
    expect(queryByText(/undefined/)).toBe(null);
  });

  test('array of error messages should all be rendered', () => {
    const { queryByText } = render(
      <FormField errorMessage={['A first error', 'A second error', undefined]}>
        Form elements here
      </FormField>,
    );
    expect(queryByText(/A second error/)).not.toBe(null);
    expect(queryByText(/A first error/)).not.toBe(null);
    expect(queryByText(/undefined/)).toBe(null);
  });

  test('hint should be rendered', () => {
    const { queryByText } = render(
      <FormField hint={<span>You should look here</span>}>Form elements here</FormField>,
    );
    expect(queryByText(/You should look here/)).not.toBe(null);
  });

  test('className prop', () => {
    const className = 'center';
    const { getByText } = render(<FormField className={className}>Form elements here</FormField>);
    const formFieldElement = getByText(/Form elements here/)!;

    const renderedClassNames = formFieldElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });
});
