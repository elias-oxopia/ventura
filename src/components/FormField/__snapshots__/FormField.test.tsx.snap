// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField isSpaced prop 1`] = `
<DocumentFragment>
  <div
    class="ventura isSpaced"
  >
    Form elements here
  </div>
</DocumentFragment>
`;

exports[`FormField snapshot 1`] = `
<DocumentFragment>
  <div
    class="ventura"
  >
    Form elements here
  </div>
</DocumentFragment>
`;
